In [1]:
#基礎課題15.1
while False:
    print('loop')

In [2]:
#基礎課題15.2
import numpy as np
n = 5
goukei = 1
for i in np.arange(1, n+1):
    goukei = goukei * i
    
print(goukei)

120


In [3]:
# 発展課題15.3
import numpy as np
lose_count =  0 # 読者が勝利した際のカウント
while lose_count <= 4 : 
    my_hand = input('my_handの入力: rock, scissors, or, paper : ')
    cpu_select = np.random.randint(3)
    if cpu_select == 0:
        cpu_hand = 'paper'
    elif cpu_select == 1:
        cpu_hand = 'scissors'
    else:
        cpu_hand = 'rock'    
    print('my_hand is ', my_hand, '. cpu_hand is ', cpu_hand)
    if my_hand == cpu_hand:
        print('***rock, paper, scissors***')
    elif my_hand == 'rock' and cpu_hand == 'scissors':
        lose_count +=1 # コンピュータが敗北した回数を加算
        print('***You win. lose_count', lose_count ,'***')
    elif my_hand == 'scissors' and cpu_hand == 'paper':
        lose_count +=1 # コンピュータが敗北した回数を加算
        print('***You win. lose_count', lose_count ,'***')
    elif my_hand == 'paper' and cpu_hand == 'rock':
        lose_count +=1 # コンピュータが敗北した回数を加算
        print('***You win. lose_count', lose_count ,'***')        
    else:
        print('***You lose.***')
print('*******Game Over*******')

my_handの入力: rock, scissors, or, paper : scissors
my_hand is  scissors . cpu_hand is  scissors
***rock, paper, scissors***
my_handの入力: rock, scissors, or, paper : scissors
my_hand is  scissors . cpu_hand is  rock
***You lose.***
my_handの入力: rock, scissors, or, paper : scissors
my_hand is  scissors . cpu_hand is  rock
***You lose.***
my_handの入力: rock, scissors, or, paper : scissors
my_hand is  scissors . cpu_hand is  paper
***You win. lose_count 1 ***
my_handの入力: rock, scissors, or, paper : scissors
my_hand is  scissors . cpu_hand is  scissors
***rock, paper, scissors***
my_handの入力: rock, scissors, or, paper : scissors
my_hand is  scissors . cpu_hand is  paper
***You win. lose_count 2 ***
my_handの入力: rock, scissors, or, paper : scissors
my_hand is  scissors . cpu_hand is  paper
***You win. lose_count 3 ***
my_handの入力: rock, scissors, or, paper : scissors
my_hand is  scissors . cpu_hand is  scissors
***rock, paper, scissors***
my_handの入力: rock, scissors, or, paper : scissors
my_hand is  s

In [4]:
# 発展課題15.4

In [4]:
from turtle import *
import numpy as np
import time 
speed(0) # 亀を高速に移動 
shapesize(0.3) # ペンサイズを小さめに設定
penup() # 亀の足跡だけを打てるように準備
np.random.seed(seed = 1) # 再現可能なように設定
in_c = 0 # 円の内部に点をプロットした回数
out_c = 0 # 円の外部に点をプロットした回数
# シミュレーションの初期設定
error = 1000000 # 誤差計算用変数：誤差を大きく設定
epsilon = 1 # 誤差の許容範囲を設定する変数 
start = time.time() # プログラムの開始時刻を保存

# ヒント：繰り返しの条件式には変数errorとepsilonを利用
while error > epsilon: # epsilonを超える誤差の間繰り返す
  x = np.random.randint(-100, 100) # 横軸の乱数を設定
  y = np.random.randint(-100, 100) # 縦軸の乱数を設定
  goto(x, y) # 亀がxとyの位置に移動する
  # 三平方の定理を利用して、以下の条件式が正しければ円内にプロットしたとする
  if -10000 <= (x * x + y * y) <= 10000: # 三平方の定理
    pencolor('red') # 円内のため赤
    shape('circle') # ペン先を円に
    stamp() # 赤いスタンプを押す
    in_c += 1 # ヒント：円内に点をプロットした数として変数に加算
  else: # 縦横100の正方形内だが円の外にプロットする
    pencolor('blue') # 円外のため青
    shape('square') # ペン先を四角に
    stamp() # 青いスタンプを押す
    out_c += 1 # ヒント：円外に点をプロットした数として変数に加算
  # 正確な円周率 - 計算中の円周率の絶対値を誤差とする
  error = abs(np.pi - in_c / (in_c + out_c) * 4) # ヒント:空欄は円周率の近似
done() # 計算終了。以下で計算結果をまとめて出力
print('円周率は',in_c / (in_c + out_c) * 4)
print('亀は', (in_c + out_c), '回動きました。')
print('亀は計算に', time.time() - start, '秒かかりました。')

円周率は 4.0
亀は 1 回動きました。
亀は計算に 1.132751226425171 秒かかりました。


In [6]:
from turtle import *
import numpy as np
import time 
np.random.seed(seed=1)
start = time.time()
speed(0) #亀を高速に移動 
in_c = 0 #  円の内部に点を打った回数
out_c = 0 # 円の外部に点を打った回数
penup() #亀の足跡だけを打てるように準備
error = 1000000
epsilon = 0.001 # 誤差の許容範囲 
shapesize(0.3)
while error > epsilon: # 誤差がepsilon以上の間
  # x軸の数値をランダムに設定 
  x = np.random.randint(-100,100)
  # y軸の数値をランダムに設定 
  y = np.random.randint(-100,100)
  goto(x, y) # 亀がxとyの位置に移動する

  if -10000 <= (x*x + y*y) <= 10000:
    pencolor('red') # 円内のため赤
    shape('circle') #  ペン先を円に
    stamp()     #dot(5) #亀の足を使いスタンプ
    in_c += 1 #円内に点が打たれた数
  else:
    pencolor('blue') #円内のため青
    shape('square') # ペン先を四角に
    stamp()     #dot(5) #亀の足を使いスタンプ
    out_c += 1 #円外に点が打たれた数
  error = abs(np.pi - in_c / (in_c + out_c) * 4)
print('円周率は',in_c / (in_c + out_c) * 4)
print('亀は', (in_c + out_c) , '回動きました。')
print('亀は計算に', time.time() - start, '秒かかりました。')
done()# 亀のお仕事終了



円周率は 3.141141141141141
亀は 666 回動きました。
亀は計算に 26.6227388381958 秒かかりました。


# 問1と問2の答え

以下の結果は実行ごとに変化することがあります。

- epsilon = 1
   - 円周率4.0 
   - 処理時間　0.7481439113616943
- epsilon = 0.1
   - 円周率3.2 
   - 処理時間　0.9422028064727783
- epsilon = 0.01
   - 円周率3.142857142857143 
   - 処理時間　1.015904188156128
- epsilon = 0.001
   - 円周率3.141141141141141
   - 処理時間　32.11669111251831

epsilonの値が低くなれば、誤差が小さくなるので、その誤差以下まで繰り返すため、繰り返しの回数が増える。

In [1]:
import numpy as np
print(np.pi)

3.141592653589793
